pkg/controller/istiocontrolplane/inputsv2.go

```diff
@@ -0,0 +1,43 @@
+package istiocontrolplane
+
+import (
+	"istio.io/operator/pkg/apis/istio/v1alpha2"
+	"istio.io/operator/pkg/helmreconciler"
+)
+
+// IstioRenderingInput is a RenderingInput specific to an v1alpha2 IstioControlPlane instance.
+type IstioRenderingInputV2 struct {
+	instance *v1alpha2.IstioControlPlane
+	crPath   string
+}
+
+// NewIstioRenderingInput creates a new IstioRenderiongInput for the specified instance.
+func NewIstioRenderingInputV2(instance *v1alpha2.IstioControlPlane) *IstioRenderingInputV2 {
+	return &IstioRenderingInputV2{instance: instance}
+}
+
+// GetCRPath returns the path of IstioControlPlane CR.
+func (i *IstioRenderingInputV2) GetCRPath() string {
+	return i.crPath
+}
+
+// GetProcessingOrder returns the order in which the rendered charts should be processed.
+func (i *IstioRenderingInputV2) GetProcessingOrder(manifests helmreconciler.ChartManifestsMap) ([]string, error) {
```

Comment:
I'm not sure if this is sufficient to express dependency relationships. Right now we have crds, which have to complete, then everything else, which can be done in parallel. 
I think what you have here implies that we serialize everything, which will make the apply run slower.

Reply:
fwiw, I don't think it is a grand idea to serialize.  I also don't think its a grand idea to run all of the chart installs in parallel.  While the iniital apply may be faster, what we are after is a faster overall startup experience.  The ordering should be something like
1. crds
1. galley
1. citadel (galley and citadel may need to be swapped)
1. everything else can likely be started in parallel after the above ordering is met.

Not sure where CNI fits in the ordering.  where is defaultProcessingOrder deifned?

I'd recommend a list of lists for this purpose.

Reply:
CRDs should probably be managed as part of the operator install.  When using Operator Lifecycle Manager (OLM), OLM manages installation of all CRDs associated with an operator.  This prevents multiple operators that manage the same resource types from being installed at the same time, while also allowing for multiple instances of the operator (depending on which namespaces the operator is configured to manage).

FWIW, in our testing on Maistra, full install takes around two minutes, which includes waiting for deployments of each component to become available before proceeding to subsequent components.  Parallel install is much faster, but there is also a settling period.

Reply:
That said, it's unclear whether or not the operator should claim the resources used by the Istio components themselves (we've been debating that ourselves).

Reply:
I think I agree with CRDs should be installed as part of operator install.  This is pretty much what the operator framework recommends in their sample operator.  @richardwxn I assume you are not installing crds per this code because crds have its own helm chart anyway.

Reply:
@sdake that is defined in inputs.go, currently set as ```
// defaultProcessingOrder for the rendered charts
var defaultProcessingOrder = []string{
	"istio",
	"istio/charts/security",
	"istio/charts/galley",
	"istio/charts/prometheus",
	"istio/charts/mixer",
	"istio/charts/pilot",
	"istio/charts/gateways",
	"istio/charts/sidecarInjectorWebhook",
	"istio/charts/grafana",
	"istio/charts/tracing",
	"istio/charts/kiali",
}
```

Reply:
@linsun @rcernich we don't want to make OLM mandatory and what we have today handles  completely arbitrary dependency relationships (e.g. all wait for CRDs, mixer and policy wait for galley): https://github.com/istio/operator/blob/master/pkg/manifest/installer.go#L103
I don't see any reason to complicate matters by treating CRDs as anything other that just another component.

Reply:
CRDs are global and you only apply once per cluster... and nothing specific to namespace.  Basically you'd run through that "applyall" code with every istio control plane resource change including CRDs.

Reply:
> @linsun @rcernich we don't want to make OLM mandatory and what we have today handles completely arbitrary dependency relationships (e.g. all wait for CRDs, mixer and policy wait for galley): https://github.com/istio/operator/blob/master/pkg/manifest/installer.go#L103
> I don't see any reason to complicate matters by treating CRDs as anything other that just another component.

@ostromart I am not clear how OLM was conflated with ordering.  I'd like to clear a few things up so we can avoid confusion in the future.

1. OLM is a framework for managing the operator lifecycle.  It simply requires a controller that operates based upon changes to a CRD (IstioControlPlane).  Nothing more.
1. The discussion about operator-sdk is not relevant here.  As far as I can tell, there is no strong motivation for Istio to take a hard dependency on operator-sdk.
1. OLM and operator-sdk are two distinct technologies.
1. OLM consumes an operator, therefore, Istio's operator would never take a hard dependency on OLM.  Instead, OLM would take a (soft) dependency on the Istio operator.  
1. All OLM requires is 1) a bunch of metadata 2) an operator that functions properly.
1. I recommend a focus on _merging_ a controller that functions properly and worry about OLM after we have a functional controller in the tree.

As for ordering - this is critical to get right, and I'm not clear this ordering approach is correct.  We can iterate independently on improvements once this PR merges rather than getting jammed up in a debate about how the implementation should be.

Reply:
> @sdake that is defined in inputs.go, currently set as
> 
> ```
> // defaultProcessingOrder for the rendered charts
> var defaultProcessingOrder = []string{
> 	"istio",
> 	"istio/charts/security",
> 	"istio/charts/galley",
> 	"istio/charts/prometheus",
> 	"istio/charts/mixer",
> 	"istio/charts/pilot",
> 	"istio/charts/gateways",
> 	"istio/charts/sidecarInjectorWebhook",
> 	"istio/charts/grafana",
> 	"istio/charts/tracing",
> 	"istio/charts/kiali",
> }
> ```
Thanks @linsun. I'll wait for this PR to merge, then take on sorting out a proper start order and synchronization.  The ordering should be (off top of my head):

1. CRDs
1. security
1. galley
1. sidecar-injector
1. pilot, mixer, gateways, prometheus
1. grafana, tracing, kiali

In the meantime, @richardwxn can you get moving on this PR?  At this pace, we will have to pull the operator from our 1.4 roadmap...

Reply:
Also @richardwxn @ostromart @sdake I think there are 2 types of CRDs and let's not confuse with them.  1) CRD is the istio control plane CRD (do we have other CRDs??) used by the operator.  2) CRDs used by istio control plane, we currently have ~20 CRDs or so.  For 2, I agree, it should be handled as the rest of Istio components.  For 1, I think the istio control plane CRD should be installed whenever the istio operator is installed.

Reply:
@linsun, the latest code does exactly as you say. The Istio CRDs are treated as a base component by the operator and all other components have a dependency on it. Beyond that, I have all other components as siblings - the relationship is defined in inputs.go componentDependencies and applied in reconciler.go processRecursive(). But that can be changed if we desire a more complex dependency tree (e.g. wait for citadel etc.).
The operator itself is installed using a bootstrap manifest (see ./build and ./deploy directories) and uses the MeshConfig CR, which is installed alongside it.